### DISCUSSION TITLE
Questions about Prisma and your Migration

### DISCUSSION THREAD

**Discussion by rntjr** [prisma/prisma#3527] [Q&A] [answered]
Hello Prisma Team. I am migrating from Java EE to Node JS. I used Spring Boot as a framework and I didn't have any difficulties in migrating to Typescript.
I was looking for a great Migration tool to deal with databases on large scales (around 3000 tables, 3000 functions, more than 1000 triggers and so on). I found your Migrate and I really liked the Introspection, but I haven't tested it with my base yet.
I would like to know from the team if the tool has any way of doing Migrations with Functions, Procedures, Triggers and the like.
Even though it is something like Flyway, it would already be a great help for technology migration.
Our product has many reports and an integrated BI, where all data is consulted and assembled through these Functions and the Triggers in some CRUDS that depend on registering some data in a table when there is a specific Insert.

**Comment by ryands17** [comment]
Hey @rntjr 👋 
Currently Prisma's introspection will only take into account the number of tables that you have. Functions and triggers will not be introspected. You can read more about introspection [here](https://www.prisma.io/docs/reference/tools-and-interfaces/introspection#overview).

After the tables are introspected, you can directly query for them via Prisma after running `prisma generate`. So that will work out of the box.

As a workaround for functions, you can use `prisma.queryRaw` to call functions with the help of a Raw Query.
Let me know if this helps :)

**Comment by rntjr** [comment, marked as answer]
Is there any way, by Prisma itself, to create a CREATE FUNCTION and CREATE TRIGGER with executeRaw?

**Reply by ryands17** [reply]
Yes you can do that via `prisma.executeRaw`. It will accept any raw SQL like creating functions and triggers.

**Reply by gdarquie** [reply]
In fact, not really, 

It is **not** possible to execute all raw commands sql  with primsa, only some of them, excluding the triggers, even with executeRaw and executeUnsafe because prisma use the binary protocol. It is a know issue but not prioritized. More info here: https://github.com/prisma/docs/issues/716

For example, if you try to create a trigger  on MariaDB with 
```js
const result = await staticClient.$executeRaw`CREATE TRIGGER test_trigger AFTER INSERT ON clients FOR EACH ROW INSERT INTO clients SET name='pseudo', reference='ref', token='ref1'`;
```

it should return the following message error 'This command is not supported in the prepared statement protocol yet'.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
